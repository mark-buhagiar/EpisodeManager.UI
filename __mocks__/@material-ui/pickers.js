import React from 'react';

const actual = jest.requireActual('@material-ui/pickers');
const moment = jest.requireActual('moment');

export default {
    __esModule: true,
    ...actual,
    DatePicker: () => (props) => {
        return (
            <input
                data-testid="mockedDateField"
                onChange={(e) => {
                    props.onChange(moment(e.target.value));
                }}
            />
        );
    },
};
