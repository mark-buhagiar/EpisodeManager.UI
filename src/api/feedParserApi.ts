import axios from 'axios';
import ApiConfig from '../config/apiConfig';
import ParseHistoryListing from '../models/ParseHistoryListing';
import ParseHistoryDetails from '../models/ParseHistoryDetails';
import moment from 'moment';

const buildEndpoint = (endpoint: string): string => ApiConfig.feedParserApiBase + endpoint;

const feedParserMSEndpoints = {
    syncWithShowRss: buildEndpoint('sync/SyncWithShowRss'),
    forceSyncParseHistory: buildEndpoint('sync/ForceSyncParseHistory'),
    getParseHistoryBetweenDates: buildEndpoint('ParseHistory/GetParseHistoryBetweenDates'),
    getParseHistoryDetails: buildEndpoint('ParseHistory/GetParseHistoryDetails'),
};

export async function syncWithShowRss(): Promise<void> {
    try {
        await axios.post(feedParserMSEndpoints.syncWithShowRss);
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}

function convertDates(listing: ParseHistoryListing): ParseHistoryListing {
    return {
        ...listing,
        retrievedOn: moment.utc(listing.retrievedOn).toDate(),
        lastSyncedOn: moment.utc(listing.lastSyncedOn).toDate(),
    };
}

export async function getParseHistoryBetweenDates(startDate: Date, endDate: Date): Promise<ParseHistoryListing[]> {
    try {
        const params = {
            startDate,
            endDate,
        };
        const parseHistory = await axios.get<ParseHistoryListing[]>(feedParserMSEndpoints.getParseHistoryBetweenDates, {
            params,
        });

        return parseHistory.data.map(convertDates);
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}

export async function getParseHistoryDetails(id: number): Promise<ParseHistoryDetails> {
    try {
        const params = {
            parseHistoryId: id,
        };
        const parseHistoryDetails = await axios.get<ParseHistoryDetails>(feedParserMSEndpoints.getParseHistoryDetails, {
            params,
        });

        return parseHistoryDetails.data;
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}
export async function forceSyncParseHistory(id: number): Promise<void> {
    try {
        const params = {
            parseHistoryId: id,
        };

        await axios.post(feedParserMSEndpoints.forceSyncParseHistory, null, {
            params,
        });
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}
