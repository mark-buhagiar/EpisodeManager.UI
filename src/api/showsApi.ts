import axios from 'axios';
import Episode from '../models/Episode';
import ApiConfig from '../config/apiConfig';

const buildEndpoint = (endpoint: string): string => ApiConfig.showsApiBase + endpoint;

const showsMSEndpoints = {
    getForCurrentUserBetweenDates: buildEndpoint('episodes/getForCurrentUserBetweenDates'),
};

export async function getForCurrentUserBetweenDates(startDate: Date, endDate: Date): Promise<Episode[]> {
    try {
        const params = {
            startDate: startDate,
            endDate: endDate,
        };
        const result = await axios.get<Episode[]>(showsMSEndpoints.getForCurrentUserBetweenDates, { params });
        return result.data;
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}
