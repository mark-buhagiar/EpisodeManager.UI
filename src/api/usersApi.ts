import axios from 'axios';
import ApiConfig from '../config/apiConfig';
import EpisodePreferences from '../models/EpisodePreferences';
import Show from '../models/Show';

const buildEndpoint = (endpoint: string): string => ApiConfig.usersApiBase + endpoint;

const usersMSEndpoints = {
    addEpisodeDownloaded: buildEndpoint('subscriptions/AddEpisodeDownloaded'),
    getUserSubscriptionsForListing: buildEndpoint('subscriptions/GetUserSubscriptionsForListing'),
    unsubscribeFromShow: buildEndpoint('subscriptions/UnsubscribeFromShow'),
    subscribeToShow: buildEndpoint('subscriptions/SubscribeToShow'),
    setUserEpisodePreferences: buildEndpoint('users/setUserEpisodePreferences'),
    getUserEpisodePreferences: buildEndpoint('users/getUserEpisodePreferences'),
};

export async function addEpisodeDownloaded(episodeId: number): Promise<void> {
    try {
        const params = {
            episodeId,
        };
        await axios.put(usersMSEndpoints.addEpisodeDownloaded, null, { params });
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}

export async function getUserSubscriptionsForListing(): Promise<Show[]> {
    try {
        const result = await axios.get(usersMSEndpoints.getUserSubscriptionsForListing);
        return result.data;
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}

export async function unsubscribeFromShow(showId: number): Promise<void> {
    try {
        const params = {
            showId,
        };
        await axios.patch(usersMSEndpoints.unsubscribeFromShow, { params });
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}

export async function subscribeToShow(showIds: number[]): Promise<void> {
    try {
        const params = {
            showIds,
        };
        await axios.put(usersMSEndpoints.subscribeToShow, null, { params });
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}

export async function setUserEpisodePreferences(preferences: EpisodePreferences): Promise<void> {
    try {
        await axios.post(usersMSEndpoints.setUserEpisodePreferences, preferences);
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}

export async function getUserEpisodePreferences(): Promise<EpisodePreferences> {
    try {
        const response = await axios.get(usersMSEndpoints.getUserEpisodePreferences);
        return response.data;
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}
