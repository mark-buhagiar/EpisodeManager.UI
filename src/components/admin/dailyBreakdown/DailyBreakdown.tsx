import moment from 'moment';
import React, { useState } from 'react';
import { Bar } from 'react-chartjs-2';
import * as showsApi from '../../../api/showsApi';
import { toDateFormat } from '../../../helpers/dateHelpers';
import { Qualities, QualityColors } from '../../../models/enums/Qualities';
import EpisodeCountDailyDistribution from '../../../models/EpisodeCountDailyDistribution';
import DateRange from '../../common/DateRange';
import Panel from '../../common/panel';
import Color from 'color';

const DailyBreakdown: React.FC = (): JSX.Element => {
    const [isLoading, setIsLoading] = useState(false);
    const [data, setData] = useState({});
    const today = new Date();

    const chartOptions = {
        scales: {
            yAxes: [
                {
                    ticks: {
                        beginAtZero: true,
                    },
                    stacked: true,
                },
            ],
            xAxes: [
                {
                    stacked: true,
                },
            ],
        },
    };

    function convertDataToChartJs(dailyDistribution: EpisodeCountDailyDistribution[]): any {
        return {
            labels: dailyDistribution.map((day) => toDateFormat(moment(day.date))),
            datasets: Qualities.map((quality) => {
                return {
                    label: quality.description,
                    backgroundColor: (QualityColors[quality.id] as Color).alpha(0.6).rgb().string(),
                    hoverBackgroundColor: (QualityColors[quality.id] as Color).alpha(0.8).rgb().string(),
                    data: dailyDistribution.map(
                        (day) => day.qualityDistribution.find((x) => x.quality === quality.id)?.count ?? 0,
                    ),
                };
            }),
        };
    }

    async function handleDateRangeChanged(startDate: Date, endDate: Date): Promise<void> {
        setIsLoading(true);
        const dailyDistribution = await showsApi.getEpisodeCountDailyDistribution(startDate, endDate);
        setData(convertDataToChartJs(dailyDistribution));
        setIsLoading(false);
    }

    const panelActions = (
        <DateRange
            initialStartDate={moment(today).subtract(6, 'days').toDate()}
            initialEndDate={today}
            onDateRangeChanged={handleDateRangeChanged}
        />
    );

    return (
        <Panel title="Daily Breakdown" panelActions={panelActions}>
            <Bar data={data} options={chartOptions} />
        </Panel>
    );
};

export default DailyBreakdown;
