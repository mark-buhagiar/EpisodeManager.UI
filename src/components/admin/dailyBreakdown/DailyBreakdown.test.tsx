import React from 'react';
import { render, fireEvent, act } from '@testing-library/react';

import DailyBreakdown from './DailyBreakdown';
import { getEpisodeCountDailyDistribution } from '../../../api/showsApi';
import { DatePicker } from '@material-ui/pickers';
import moment from 'moment';

jest.mock('@material-ui/pickers', () => {
    const actual = jest.requireActual('@material-ui/pickers');
    const moment = jest.requireActual('moment');

    return {
        ...actual,
        DatePicker: (() => (props: any): JSX.Element => {
            return (
                <input
                    data-testid="mockedDateField"
                    onChange={(e): void => {
                        props.onChange(moment(e.target.value));
                    }}
                />
            );
        })(),
    };
});

jest.mock('react-chartjs-2', () => ({
    Bar: (): string => 'Fake Chart',
}));

jest.mock('../../../api/showsApi', () => ({ getEpisodeCountDailyDistribution: jest.fn() }));

const validResult = [
    {
        qualityDistribution: [
            {
                count: 31,
                quality: 1,
            },
            {
                count: 31,
                quality: 2,
            },
            {
                count: 27,
                quality: 3,
            },
        ],
        date: '2020-05-01T00:00:00',
    },
];

afterEach(() => {
    jest.clearAllMocks();
});

describe('When the component is loading', () => {
    it('should display a loading spinner', async (): Promise<void> => {
        jest.useFakeTimers();
        (getEpisodeCountDailyDistribution as jest.Mock).mockReturnValueOnce(Promise.resolve(validResult));
        const { asFragment } = render(<DailyBreakdown />);
        expect(asFragment()).toMatchSnapshot();

        await act(async () => {
            await Promise.resolve();
        });
    });
});

describe('When data retrieval errors out', () => {
    it('should display an error icon', async (): Promise<void> => {
        (getEpisodeCountDailyDistribution as jest.Mock).mockReturnValueOnce(Promise.reject('Error test'));
        const { asFragment } = render(<DailyBreakdown />);

        await act(async () => {
            await Promise.resolve();
        });
        expect(asFragment()).toMatchSnapshot();
    });
});

describe('When data retrieval completes successfully', () => {
    it('should display the chartjs chart', async (): Promise<void> => {
        (getEpisodeCountDailyDistribution as jest.Mock).mockResolvedValueOnce(validResult);
        const { asFragment } = render(<DailyBreakdown />);

        await act(async () => {
            await Promise.resolve();
        });
        expect(asFragment()).toMatchSnapshot();
    });
});

describe('When selected dates are changed', () => {
    it.only('should re-query the web service', async (): Promise<void> => {
        (getEpisodeCountDailyDistribution as jest.Mock).mockResolvedValueOnce(validResult);

        await act(async () => {
            const { asFragment, getAllByTestId } = render(<DailyBreakdown />);
            const mockedDateField = getAllByTestId('mockedDateField');
            fireEvent.change(mockedDateField[0], { target: { value: moment().format() } });
        });

        expect(getEpisodeCountDailyDistribution).toHaveBeenCalledTimes(2);
    });
});
