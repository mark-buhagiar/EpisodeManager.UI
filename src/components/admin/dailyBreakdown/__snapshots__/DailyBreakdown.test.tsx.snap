// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When data retrieval completes successfully should display the chartjs chart 1`] = `
<DocumentFragment>
  <div
    class="panel"
  >
    <div
      class="panel-title"
    >
      Daily Breakdown
      <div
        class="panel-actions"
      >
        <div
          class="date-range"
        >
          <input
            data-testid="mockedDateField"
          />
          <input
            data-testid="mockedDateField"
          />
        </div>
      </div>
      <div
        class="separator"
      />
    </div>
    Fake Chart
  </div>
</DocumentFragment>
`;

exports[`When data retrieval errors out should display an error icon 1`] = `
<DocumentFragment>
  <div
    class="panel"
  >
    <div
      class="panel-title"
    >
      Daily Breakdown
      <div
        class="panel-actions"
      >
        <div
          class="date-range"
        >
          <input
            data-testid="mockedDateField"
          />
          <input
            data-testid="mockedDateField"
          />
        </div>
      </div>
      <div
        class="separator"
      />
    </div>
    <div
      class="component-error-container"
    >
      <span
        class="component-error"
      >
        <span
          class="material-icons"
        >
          error_outline
        </span>
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`When the component is loading should display a loading spinner 1`] = `
<DocumentFragment>
  <div
    class="panel"
  >
    <div
      class="panel-title"
    >
      Daily Breakdown
      <div
        class="panel-actions"
      >
        <div
          class="date-range"
        >
          <input
            data-testid="mockedDateField"
          />
          <input
            data-testid="mockedDateField"
          />
        </div>
      </div>
      <div
        class="separator"
      />
    </div>
    <div
      class="component-loading-container"
    >
      <span
        class="component-loading"
      >
        <div
          class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
          role="progressbar"
          style="width: 40px; height: 40px;"
        >
          <svg
            class="MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
              cx="44"
              cy="44"
              fill="none"
              r="20.2"
              stroke-width="3.6"
            />
          </svg>
        </div>
      </span>
    </div>
  </div>
</DocumentFragment>
`;
